pkg/controller/direct/directbase/directbase_controller.go

```diff
@@ -53,22 +55,81 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
-// Add creates a new controller for reconciling objects of the specified GVK, delegating actual resource reconciliation to the provided Model.
-func Add(mgr manager.Manager, gvk schema.GroupVersionKind, model Model, opts Deps) error {
+var ControllerBuilder directControllerBuilder
+
+func init() {
+	ControllerBuilder = directControllerBuilder{}
+}
+
+type directControllerBuilder struct {
+	modelMapper map[schema.GroupVersionKind]func(*controller.Config) Model
+}
+
+func (c *directControllerBuilder) RegisterModel(gvk schema.GroupVersionKind, modelFn func(*controller.Config) Model) {
+	if c.modelMapper == nil {
+		c.modelMapper = map[schema.GroupVersionKind]func(*controller.Config) Model{}
+	}
+	c.modelMapper[gvk] = modelFn
+}
+
+func (c *directControllerBuilder) AddController(mgr manager.Manager, config *controller.Config, gvk schema.GroupVersionKind, deps Deps) error {
 	immediateReconcileRequests := make(chan event.GenericEvent, k8s.ImmediateReconcileRequestsBufferSize)
 	resourceWatcherRoutines := semaphore.NewWeighted(k8s.MaxNumResourceWatcherRoutines)
-	reconciler, err := NewReconciler(mgr, immediateReconcileRequests, resourceWatcherRoutines, gvk, model, opts.JitterGenerator)
+
+	reconciler, err := c.NewReconciler(mgr, config, immediateReconcileRequests, resourceWatcherRoutines, gvk, deps.JitterGenerator)
 	if err != nil {
 		return err
 	}
 	return add(mgr, reconciler)
 }
 
-// NewReconciler returns a new reconcile.Reconciler.
-func NewReconciler(mgr manager.Manager, immediateReconcileRequests chan event.GenericEvent, resourceWatcherRoutines *semaphore.Weighted,
-	gvk schema.GroupVersionKind, model Model, jg jitter.Generator) (*DirectReconciler, error) {
+func (c *directControllerBuilder) IsDirectByCRD(crd *apiextensions.CustomResourceDefinition) bool {
+	for gvk, _ := range c.modelMapper {
```

Comment:
I think we can just call IsDirectByGK here - from inside IsDirectByCRD (the version shouldn't matter)

Reply:
Sure. Removed `IsDirectByCRD` and `IsDirectByGVK`, and only keep `IsDirectByGK`